### PR TITLE
Switch build scripts to LDC-LLVM repo & upgrade dockerfile

### PR DESCRIPTION
**Pull request by kinke**

### CODE REVIEW

File: docker-build-release/ldc-builder-linux-x86_64/Dockerfile [hunk 1]
```diff
@@ -22,20 +22,20 @@ ENV BUILD_ROOT=/build
 RUN mkdir ${BUILD_ROOT}
 WORKDIR ${BUILD_ROOT}
 
-# The 12.04 system GCC is too old for building recent LLVM versions, so grab a
+# The 14.04 system GCC is too old for building recent LLVM versions, so grab a
 # backport from the "unofficial-official" toolchain test PPA.
-RUN apt-get -y update && apt-get -yq install python-software-properties && \
+RUN apt-get -yq install software-properties-common && \
```
Comment (dnadlinger) [inline docker-build-release/ldc-builder-linux-x86_64/Dockerfile]: Why are you removing the `update`?

Comment (kinke) [reply docker-build-release/ldc-builder-linux-x86_64/Dockerfile]: There's one further up which should make sure `software-properties-common` is found and can be installed. We do another update immediately afterwards.

Comment (dnadlinger) [reply docker-build-release/ldc-builder-linux-x86_64/Dockerfile]: The old layer might come from a cache, so might be arbitrarily old.

Comment (dnadlinger) [reply docker-build-release/ldc-builder-linux-x86_64/Dockerfile]: https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#run

Comment (kinke) [reply docker-build-release/ldc-builder-linux-x86_64/Dockerfile]: Well ending up with a somewhat outdated package for that tiny tool was the least of my concerns.

Comment (dnadlinger) [reply docker-build-release/ldc-builder-linux-x86_64/Dockerfile]: Those somewhat outdated packages are removed from mirrors quite quickly. Did you never forget to run `apt-get update` in the past and then got 404 errors? Also, not breaking things that already work should be a good starting point for any PR.

Comment (kinke) [reply docker-build-release/ldc-builder-linux-x86_64/Dockerfile]: > Also, not breaking things that already work should be a good starting point for any PR.

Regressions can always creep in as we know. And no, I don't recall any 404s, but only recently started using Linux more often. Will re-add the update, next time please just fix it yourself with an explanatory comment, that's probably more efficient if one's sure it's a regression.
Just to elaborate on my motivation: I got tired of the 3 consecutive `apt-get update` during testing.

**Comment by kinke** [comment]
Good to go? Needed for 1.4-beta1...

File: docker-build-release/ldc-builder-linux-x86_64/Dockerfile (hunk 1 above)
Comment (dnadlinger) [reply docker-build-release/ldc-builder-linux-x86_64/Dockerfile]: >  next time please just fix it yourself

I'm at a conference (in the US, hence travelling light), so I can't easily do that.

> Just to elaborate on my motivation: I got tired of the 3 consecutive apt-get update during testing.

In any sort of regular usage, the older layers would be cached, only rerunning whatever you changed (and later). This is a good part of the power of Docker.

**Comment by dnadlinger** [comment]
I've merged this, but is there a good reason to go away from 12.04? We are still getting questions from people (HPC and similar environments) where even the 12.04 glibc is still too new. There are different solutions for this, of course (e.g. the Guix-based redistributable packages), but as long as there is no pressing reason to break backwards compatibility, it doesn't seem wise to do so.

**Comment by kinke** [comment]
Well 12.04 has reached its end of life and no common CI system doesn't support it anymore, so I thought it'd just be logical to move on to 14.04. My secret ;) hope was that this may enable LTO out-of-the-box for most Debian-based users (our `LLVMgold.so` hopefully being compatible with the default system `ld`), but I haven't investigated further.

**Comment by dnadlinger** [comment]
> so I thought it'd just be logical to move on to 14.04

The whole point of using an old distro for this was backwards-compatibility. What CI systems use doesn't enter that consideration (well, apart from placing an upper limit on the version we use).

**Comment by kinke** [comment]
To me backwards compatibility ends where our prebuilt packages might be linked against vulnerable old libraries not being officially supported & updated anymore.

**Comment by wilzbach** [comment]
> We are still getting questions from people (HPC and similar environments) where even the 12.04 glibc is still too new.

FWIW I have run into this problem in the past as well. Of course building from source isn't too difficult, but I also see the point about not using a non-supported version anymore.

**Pull request merged**